Applications/Browser/NetworkHistoryWidget.cpp

```diff
@@ -0,0 +1,154 @@
+
+    set_layout<GUI::VerticalBoxLayout>();
+    layout()->set_margins({ 4, 4, 4, 4 });
+    layout()->set_spacing(4);
+
+    auto& toolbar_container = add<GUI::ToolBarContainer>();
+    auto& toolbar = toolbar_container.add<GUI::ToolBar>();
+
+    auto& paused = toolbar.add<GUI::CheckBox>("Paused");
+    paused.on_checked = [this](bool checked) { m_paused = checked; };
+    paused.set_checked(m_paused);
+    paused.set_size_policy(GUI::SizePolicy::Fixed, GUI::SizePolicy::Fixed);
+    paused.set_preferred_size(61, 26);
+
+    auto& clear_on_navigation = toolbar.add<GUI::CheckBox>("Clear on navigation");
+    clear_on_navigation.on_checked = [this](bool checked) { m_clear_on_navigation = checked; };
+    clear_on_navigation.set_checked(m_clear_on_navigation);
+    clear_on_navigation.set_size_policy(GUI::SizePolicy::Fixed, GUI::SizePolicy::Fixed);
+    clear_on_navigation.set_preferred_size(131, 26);
+
+    auto& disable_cache = toolbar.add<GUI::CheckBox>("Disable cache (while open)");
+    disable_cache.on_checked = [this](bool checked) { m_cache_disabled = checked; };
+    disable_cache.set_checked(m_cache_disabled);
+    disable_cache.set_size_policy(GUI::SizePolicy::Fixed, GUI::SizePolicy::Fixed);
+    disable_cache.set_preferred_size(166, 26);
+
+    auto& splitter = add<GUI::VerticalSplitter>();
+
+    m_history_view = splitter.add<GUI::TableView>();
+    m_history_view->set_model(Web::NetworkHistoryModel::create(m_network_history));
+
+    // Path can be very long, so it's hidden by default.
+    m_history_view->set_column_hidden(Web::NetworkHistoryModel::Column::Path, true);
+
+    m_context_menu = GUI::Menu::construct();
+
+    m_context_menu->add_action(GUI::Action::create("Open in new tab", [this](auto&) {
+        if (on_tab_open_request)
+            on_tab_open_request(m_context_menu_url);
+    }));
+
+    m_history_view->on_context_menu_request = [this](const auto& index, const auto& event) {
+        if (!index.is_valid())
+            return;
+
+        m_context_menu_url = m_history_view->model()->data(index, GUI::Model::Role::Custom).to_string();
+        m_context_menu->popup(event.screen_position());
+    };
+}
+
+NetworkHistoryWidget::~NetworkHistoryWidget()
+{
+    unregister_callbacks();
+}
+
+void NetworkHistoryWidget::update_view()
+{
+    m_history_view->model()->update();
+}
+
+void NetworkHistoryWidget::register_callbacks()
+{
+    // FIXME: Resource loader can't tell which tab the load came from,
+    //        so all tabs will report to the most recently opened window.
+    Web::ResourceLoader::the().on_load = [this](auto load_id, auto& url) {
+        if (m_paused)
+            return;
+
+        Web::NetworkHistoryModel::Entry new_entry;
+        new_entry.url = url;
+        new_entry.load_timer.start();
+
+        m_network_history.set(load_id, new_entry);
+
+        // This will add some microseconds to the load timer, which isn't that big of a deal.
+        update_view();
```

Comment:
Guess you could schedule the update for later, instead of doing it inside the callback, if that's a concern

Reply:
Eh, it's only really noticable with cached resources, where it switches between 0 and 1 ms. So currently it's not much of a concern.